Added Issue reporting and triage process document

vikramkoka opened the pull request:
Added Issue reporting and triage process document, including the
updated set of labels to be used for reporting and tracking issues.


---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

vikramkoka committed:
Added Issue reporting and triage process document
Added Issue reporting and triage process document, including the
updated set of labels to be used for reporting and tracking issues.

jhtimmins commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+
+This document explains the issue tracking and triage process within Apache
+Airflow including labels, milestones, and priorities as well as the process
+of resolution of issues.
```
```suggestion
of resolving issues.
```

vikramkoka replied:
done

jhtimmins commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+
+**Kind**
+
+The “kind” of labels indicate “what kind of issue it is”. The most
```
```suggestion
The “kind” labels indicate “what kind of issue it is”. The most
```

vikramkoka replied:
done

jhtimmins commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+change to be done as part of an AIP or something which needs to be cleaned
+up in the project.
+
+Issues of ``kind:documentation`` are exactly that, changes which need to be
```
```suggestion
Issues of ``kind:documentation`` are for changes which need to be
```

vikramkoka replied:
done

jhtimmins commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+being released and versioned independently.
+
+There are more detailed areas of the project such as Scheduler, Webserver,
+API, UI, Logging, and Kubernetes which are all conceptually under the
```
```suggestion
API, UI, Logging, and Kubernetes, which are all conceptually under the
```

jhtimmins commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+
+Therefore, the priority labels used are:
+
+* ``Priority:critical``: Showstopper bug that should be resolved immediately and a patch issued as soon as possible. Typically, this is because it affects most users and would take down production systems.
```
Does it make sense to capitalize the `Priority` prefix if we don't capitalize `area`?

vikramkoka replied:
You are right, it should have been lower case. Fixed it.

jhtimmins commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+
+The importance of using priority effectively is to both triage incoming issues
+appropriately as well to make sure that when we release a new version of Airflow,
+we can ship a release confident that there are no “production blocker” issues in it.
```
```suggestion
It's important to use priority labels effectively so we can triage incoming issues
appropriately and make sure that when we release a new version of Airflow,
we can ship a release confident that there are no “production blocker” issues in it.
```

mik-laj commented on docs/apache-airflow/index.rst:
```diff
@@ -104,6 +104,7 @@ unit of work and continuity.
     production-deployment
     backport-providers
     faq
+    Issue reporting and resolution process <issue-triage-process>
```
Is this documentation intended for the end user or for a contributor? I think it would be better if this link got added to CONTRIBUTING.rst.

vikramkoka replied:
I believe it is useful for both end users who report issues and want to understand the process of "what next", as well as for contributors. 
I do think adding this link to the CONTRIBUTING.rst also makes sense, once this is reviewed and published.

mik-laj replied:
how often do these rules change?  as often as a new release comes out, or more often? Is it related to specific Airflow version?

vikramkoka replied:
I am sorry, I don't understand the question.

mik-laj replied:
The apache-airflow package is versioned according to the apache-airflow release cycle. If you think it is worthwhile for users to have access to this document, we need to consider whether users will have access to a good version of this document. It seems to me that this page is not related to any particular version of apache-airflow and will therefore be out of date very quickly. It may happen that after the first day of release of a new version of apache-airflow, we will want to update the process a bit, e.g. to add new labels, but this will be a bit difficult because we try to keep the documentation immutable. We only update the documentation when a new version is released.

Please note where this page is posted on the website. If you want to go to this article, you will have to go to airflow.apache.org, click "Documentation", click "Apache Airflow", and. finally click "Issue reporting and resolution process". If you want to post it on airflow.apache.org then it's a better idea to add it to the section ["Community"](http://airflow.apache.org/community/), which already describes the rules for reporting bugs.

potiuk replied:
I agree with @mik-laj. This document is likely to change between releases, so it falls into the same category as CONTRIBUTING.rst or  README. I agree that it should be in GitHub rather than in the generated documentation. The best we can do is to have it in GitHub and link to it from the documentation if we want to make users aware of its existence. Everything it talks about is GitHub issues, so there is no problem with it being in GitHub and not under `airflow.apache.org`.

vikramkoka replied:
Got it. 
I removed the change to the documentation index and moved this new .rst file to the top level directory.
I added a reference to it in CONTRIBUTING.rst

vikramkoka committed:
Fixed grammatical errors
Fixed grammatical errors based on feedback.

vikramkoka committed:
Fixed case in priority
Changed the priority word in labels to lower case.

potiuk commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+a release. Once a PR is created to fix that issue and when that PR is tagged with a
+milestone, it implies that the PR is intended to released in that milestone.
+
+It is worth noting that Airflow Core and Airflow Providers are now released and
```
I am not sure if we want to actually have milestones for Providers. We could have a separate milestone for every batch of providers for sure, but I think this is not necessary, as we will always release whatever we have in master as the next wave of providers. What is more important is a list of upcoming changes in providers - so an automated way to show what will be included in the next wave of providers (so everything that is merged already) - and we will have a separate issue which will get updated automatically with:

* list of providers that we will release (i.e. all providers that have some changes)
* for each provider, detailed list of changes that we will release

This list will be automatically updated and once we release a next wave of providers, it will clean-up automatically.

I believe this is a bit different than core Airflow. For Airflow we deliberately choose what we are going to release by cherry-picking (at least for patchlevels), where for Providers we always release everything we have in master.

vikramkoka replied:
Hmm, that's a good point. 
I was basing this on the behavior to date including the milestone "apache-airflow-providers-google v2" which is currently there in github issues. 
I think you are making the point, that going forward this should be an exception rather than the rule. 
Did I understand you correctly?

potiuk replied:
Correct. I am planning to propose how formally our provider release process will look like (basically release all provider that changed from master every 3 weeks in regular intervals (RC - 1 week vote  - Release - 2 weeks break) - i wanted to discuss it at the dev call today. and in this case we can set a release schedule in calendar rather than a milestone.

vikramkoka replied:
Done. I explicitly made the milestone reference only towards Airflow Core releases now. Can update this later once the Provider release proposal is locked down.

eladkal commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+Similarly, some providers are broken out such as Apache (``provider/Apache``),
+AWS (``provider/AWS``), Azure (``provider/Azure``), and Google (``provider/Google``).
+These make it easier for developers working on a single provider to
+track issues for that provider.
```
Should we mention the motivation why some providers have dedicated labels while other don't?

vikramkoka replied:
done

eladkal commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+but in practice most “feature requests” are actually enhancement requests,
+so we decided to combine them both into ``kind:feature``.
+
+We recently introduced ``kind:task`` to categorize issues which are
```
```suggestion
The ``kind:task`` is used to categorize issues which are
```
i'm not sure if it matters that this was recently added? Also, saves the future issue of fixing the sentence when "recently" is not true any longer.

vikramkoka replied:
Done.

eladkal commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+API, UI, Logging, and Kubernetes, which are all conceptually under the
+“Airflow Core” area of the project.
+
+Similarly, some providers are broken out such as Apache (``provider/Apache``),
```
Is broken out the proper way to describe it? I'm not native English speaker but this sounds a bit odd to me. Maybe split / divide are more suitable?

vikramkoka replied:
tried again, please let me know if this is better.

eladkal commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+
+It's important to use priority labels effectively so we can triage incoming issues
+appropriately and make sure that when we release a new version of Airflow,
+we can ship a release confident that there are no “production blocker” issues in it.
```
We need to clarify if the intention to use priority labels only on Airflow core or also on providers.
in general issue can be critical in a specific provider (broken provider, broken specific package in provider etc..)
If we set `priority:critical` on Google provider. Does this means we expect/suggest/consider a patch release for this provider?

potiuk replied:
I believe yeah, this might trigger us to have an out-of-band release in case we find a critical issue - especially that we can do it easily now and nothing stops us except the efficiency/batching of voting process.

vikramkoka replied:
Thank you, great point. done.

eladkal commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+
+
+
+**Invalidity**
```
we should also add note about `won't fix`

vikramkoka replied:
done

eladkal commented on docs/apache-airflow/issue-triage-process.rst:
```diff
@@ -0,0 +1,195 @@
+
+This document explains the issue tracking and triage process within Apache
+Airflow including labels, milestones, and priorities as well as the process
+of resolving issues.
```
Should we mention here that there is no need to open an issue prior to opening a PR?

potiuk replied:
Agree. Good point. People do not realise that -it's a rare thing not having to open issue. I think we are pioneers with that. i actually find it pretty cool that we base all that we do on PR numbers rather than issue numbers. This is sooo much less overhead, but people do not realise that this is the case.

Actually we could even modify our template for new issues to include that information alongside "Please save some trees and do not create issue if you can just open PR fixing the problem" ;).

vikramkoka replied:
Good point, done.

eladkal reviewed (changes requested):
Great summary!

vikramkoka committed:
Updated issue triage document based on feedback
Updated the issue reporting and triage document based on the feedback
received on the PR.

vikramkoka committed:
Moved the issue process doc
Based on PR feedback, moved the reference to the issue process doc from the
airflow docs to Contributing doc. Also moved the .rst file to the right directory
and renamed it to match convention.
